### ISSUE TITLE
Higher quality icons

### ISSUE THREAD

**Issue by yuvalkarmi** [paymoapp/node-active-window#6]
Hi, thanks so much for creating this great project!

I'm wondering if there's a way to get a higher quality version of the icon. Currently testing this on a mac, but this is relevant cross-platform.

Digging through the source code, I did notice that the linux implementation does specify three sizes. I'm not sure how that's used / what's returned, but it would be super useful to be able to specify quality / pixel size desired.

However, it is worth noting that I am not using the native libraries directly, but rather the node binding :)

Thank you!

**Comment by gergof** [comment]
Hi!

The icon size currently fetched follows this logic:

##### MacOS
We use [NSRunningApplication.icon](https://developer.apple.com/documentation/appkit/nsrunningapplication/1529885-icon?language=objc) which returns an NSImage and we encode that object as PNG.
The encoded image is at the full resolution of the icon. As far as I know there's no other method to extract a higher quality icon for a running application __through the API__. This doesn't require any extra permissions.
But if you are OK with having your application access the filesystem directly, you can take a look at this: https://www.howtogeek.com/687389/how-to-extract-a-high-quality-icon-from-a-mac-app/. Basically you have to open `/Applications/yourApp.app/Contents/Resources/ApplicationIcon.icns` from node and then you can parse that file to extract the high quality icons if there're any.

##### Windows
###### UWP applications
We get the folder containing the application icons and we try to fetch the icon with the format AppName.png or AppName.scale-100.png. This should return the native resolution of the icon. But for some applications higher resolution icons are available (for example scale-200, 400, etc). _Here we could get some higher quality icons_
###### Native applications
We use the executable to extract the highest resolution icon (typically 256x256 px) (https://learn.microsoft.com/en-us/windows/win32/api/shellapi/nf-shellapi-shgetimagelist#shil_jumbo-0x4).

##### Linux
This logic is the least straight forward.
At initialization time we load all the desktop entries and build a map with applicationName-iconPath pairs. The icon paths are constructed like this: `oneOfTheSupportedIconFolders/oneOfTheSupportedThemes/oneOfTheSupportedResolution/apps/iconName.png`.
This is where the array you talked about comes into place. We iterate through an array of the most popular icon resolutions. There'are higher resolution icons than 256x256 for some applications though. If a desktop entry is not available, we extract the icon directly from the X11 API, but this icon is always a lower resolution one (typically at or under 32x32 px).

In conclusion it's only trivial to get higher resolution icons on linux and on windows only for UWP apps. On Mac it involves file system access which would open up a permission popup when used.

_But_ it would be possible and in some cases helpful to explicitly set an icon size target. But it would be more like theReturnedSize<=targetSize. This could save some memory and processing power when we know that we need at most 32x32 icons for example.
